apps/prairielearn/src/models/enrollment.ts

```diff
@@ -13,6 +18,45 @@ export async function insertEnrollment({
   return await queryRow(sql.insert_enrollment, { course_instance_id, user_id }, EnrollmentSchema);
 }
 
+export async function insertCheckedEnrollment(
+  res: Response,
+  {
+    institution,
+    course_instance,
+    authz_data,
+  }: {
+    institution: Institution;
+    course_instance: CourseInstance;
+    authz_data: any;
+  },
+): Promise<boolean> {
+  // Safety check: ensure the student would otherwise have access to the course.
+  // If they don't, throw an access denied error. In most cases, this should
+  // have already been checked.
+  if (!authz_data.has_student_access) {
+    throw error.make(403, 'Access denied');
+  }
```

Comment:
This replaces the join with `check_course_instance_access` in the enrollment query in `pages/enroll/enroll.sql`.